FFTW can be specified as an option, it is also downloaded if not found

DiamonDinoia opened the pull request:
Dear all,

I made FFTW an option so that the user can pass a cmake target to choose any library he wants (e.g. intel MKL). As long as these are cmake_targets. This is particularly helpful is finufft is used in another bigger project that already uses fftw.

I also added the possibility to use CPM to download and build fftw in case it is not specified or found in the system.
Lastly I replaced some variables with options since they are command line options not variables.

Thanks,
Marco

lu1and10 commented on CMakeLists.txt:
```diff
@@ -34,23 +35,80 @@ if(FINUFFT_BUILD_MATLAB)
+        include(${CPM_DOWNLOAD_LOCATION})
+
+        set(FFTW_VERSION 3.3.10)
+        if (FINUFFT_FFTW_SUFFIX STREQUAL THREADS)
+            set(FINUFFT_USE_THREADS ON)
+        else ()
+            set(FINUFFT_USE_THREADS OFF)
+        endif ()
+        CPMAddPackage(
+                NAME fftw3
+                OPTIONS "ENABLE_AVX2 ON" "BUILD_TESTS OFF" "BUILD_SHARED_LIBS OFF" "ENABLE_THREADS ${FINUFFT_USE_THREADS}" "ENABLE_OPENMP ${FINUFFT_USE_OPENMP}"
+                URL "http://www.fftw.org/fftw-${FFTW_VERSION}.tar.gz"
+                URL_HASH "MD5=8ccbf6a5ea78a16dbc3e1306e234cc5c"
+                EXCLUDE_FROM_ALL YES
+                GIT_SHALLOW YES
+        )
+
+        CPMAddPackage(
+                NAME fftw3f
+                OPTIONS "ENABLE_AVX2 ON" "BUILD_TESTS OFF" "BUILD_SHARED_LIBS OFF" "ENABLE_FLOAT ON" "ENABLE_THREADS ${FINUFFT_USE_THREADS}" "ENABLE_OPENMP ${FINUFFT_USE_OPENMP}"
+                URL "http://www.fftw.org/fftw-${FFTW_VERSION}.tar.gz"
+                URL_HASH "MD5=8ccbf6a5ea78a16dbc3e1306e234cc5c"
+                EXCLUDE_FROM_ALL YES
+                GIT_SHALLOW YES
+        )
+        set(FINUFFT_FFTW_LIBRARIES fftw3 fftw3f)
+        if (FINUFFT_USE_THREADS)
+            list(APPEND FINUFFT_FFTW_LIBRARIES fftw3_threads fftw3f_threads)
+        elseif (FINUFFT_USE_OPENMP)
+            list(APPEND FINUFFT_FFTW_LIBRARIES fftw3_omp fftw3f_omp)
+        endif ()
+
+        foreach (element IN LISTS FINUFFT_FFTW_LIBRARIES)
+            set_property(TARGET ${element} PROPERTY POSITION_INDEPENDENT_CODE ON)
+        endforeach ()
+
+    else ()
+        set(FINUFFT_FFTW_LIBRARIES "FFTW::Float" "FFTW::Double" "FFTW::Float${FINUFFT_FFTW_SUFFIX}" "FFTW::Double${FINUFFT_FFTW_SUFFIX}")
+    endif ()
+endif ()
```
Thanks for the refinement @DiamonDinoia. One minor suggestion, this fetching fftw part may be factored to cmake directory and using include(cmake/FFTW.cmake)? BTW I mostly use native CMake FetchContent, is CPM.cmake more commonly used?

DiamonDinoia replied:
We can move this in another file if that's preferable.

Not sure what is more commonly used but CPM is a wrapper around FetchContent that makes things much easier. It is also offer the possibility of caching dependencies between multiple projects. In general, I find it more convenient. Especially, using FetchContent to specify options and exclude binaries from install is cumbersome

DiamonDinoia replied:
Actually, I think we should download the cmake module using cpm too instead of copy pasting and external file that is available on git. I will make the changes once I have time.

DiamonDinoia commented:
Dear @lu1and10 ,

I split the cmake so that the modules are in separate files.
FindFFTW has been removed from the repository,  CPM to downloads it.
If FFTW is not found is downloaded and built before finufft.

Cheers,
Marco

DiamonDinoia commented:
@ahbarnett, do you have any suggestions?

lu1and10 commented on CMakeLists.txt:
```diff
@@ -78,25 +80,25 @@ function(set_finufft_options target)
 # Main finufft libraries
 add_library(finufft_f32 OBJECT ${FINUFFT_PRECISION_DEPENDENT_SOURCES})
-target_compile_definitions(finufft_f32 PRIVATE SINGLE)
+target_compile_definitions(finufft_f32 PRIVATE -DSINGLE=1)
```
I guess we don't need define a value for SINGLE, `target_compile_definitions(finufft_f32 PRIVATE SINGLE)` should be good.

lu1and10 replied:
@DiamonDinoia I guess this is the new commits.

DiamonDinoia replied:
The change has been reverted.

DiamonDinoia replied:
The change has been reverted.

lu1and10 commented on cmake/setupFFTW.cmake:
```diff
@@ -0,0 +1,65 @@
+
+if (FINUFFT_FFTW_LIBRARIES STREQUAL DEFAULT OR FINUFFT_FFTW_LIBRARIES STREQUAL DOWNLOAD)
+    find_package(FFTW)
+    if ( (NOT FFTW_FOUND ) OR (FINUFFT_FFTW_LIBRARIES STREQUAL DOWNLOAD))
```
I guess the if in line 13 matches line 62 and line 64. While the indents in between line 14 and line 62 may be misleading.

lu1and10 replied:
@DiamonDinoia I guess this is the new commits.

DiamonDinoia replied:
The change has been reverted.

DiamonDinoia commented on CMakeLists.txt:
```diff
@@ -3,24 +3,25 @@ cmake_minimum_required(VERSION 3.19)
-set(FINUFFT_USE_OPENMP "ON" CACHE BOOL "Whether to use OpenMP for parallelization. If disabled, the finufft library will be single threaded. This does not affect the choice of FFTW library.")
-
-if(FINUFFT_BUILD_MATLAB)
+option(FINUFFT_ARCH_FLAGS "Compiler flags for specifying target architecture." -march=native)
```
I think you reviewed an old commit here.

lu1and10 replied:
Yes, this this is outdated.

DiamonDinoia commented:
I noticed an issue with the latest cmake version on Windows. Linux seemed to work fine. I could add .gitworkflows to build the code on linux using cmake.

lu1and10 commented:
> I noticed an issue with the latest cmake version on Windows. Linux seemed to work fine, I fixed it. I could add .gitworkflows to build the code on linux using cmake.

What was the issue with cmake on Windows? Could you reproduce the issue in github workflows

DiamonDinoia commented:
> > I noticed an issue with the latest cmake version on Windows. Linux seemed to work fine, I fixed it. I could add .gitworkflows to build the code on linux using cmake.
> 
> What was the issue with cmake on Windows? Could you reproduce the issue in github workflows

It is an issue with the fftw cmake file. Reproducing it? I added the public header to the fftw target created by fftw but cmake did not like it. You could checkout my previous commit. I fixed it in my latest and added a comment about the workaround.
